File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/internal/classes/invokeClassMethod.ts
import { bufferToU32, IExternalWrapInstance } from "@polywrap/reim-wrap-js";
import { WrapManifest } from '../../WrapManifest';

import { TestExternalClassWrapped } from "../../wrapped";


export function invokeClassMethod(buffer: Uint8Array, wrapInstance: IExternalWrapInstance): Promise<Uint8Array> {
  const classId = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  switch (classId) {
    
    
    
    case WrapManifest.Internal.Class.TestExternalClass:
      return TestExternalClassWrapped.invokeMethod(dataBuffer, wrapInstance);
    
    default:
      throw new Error("Unknown class ID: " + classId.toString());
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/internal/classes/invokeClassMethod.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };

use wrapped::{ TestExternalClassWrapped };


pub async fn invoke_class_method(buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
  let class_id = u32::from_be_bytes(buffer.try_into().expect("Class ID must be 4 bytes"));
  let data_buffer = &buffer[4..];

  match class_id {
    
    
    
    WrapManifest::Internal::Class::TestExternalClass =>
        TestExternalClassWrapped.invoke_method(data_buffer, external_module),
    
    _ => panic!("Unknown class ID: ".to_string() + &class_id.to_string()),
  }
}
